1710: selinux: Update the KEP for 1.33 and graduate to Beta

jsafrane opened the pull request:
- One-line PR description: Update the KEP with current development (1.32/33) and a plan to graduate to Beta in 1.33


<!-- link to the k/enhancements issue -->
- Issue link: https://github.com/kubernetes/enhancements/issues/1710

<!-- other comments or additional information -->
- Other comments: See individual small commits.
  * Specify SELinux controller tests.
  * Add notes about other Linux distros (tested on Debian).
  * Add a note that the controller / KCM cannot implement SELinux label defaulting.

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -465,13 +469,13 @@ spec:
 
 **Feature gates `SELinuxMountReadWriteOncePod == true` && `SELinuxMount == false`**:
-* If `myclaim` is a RWOP volume (`Spec.AccessModes == ["ReadWriteOncePod']`)  *and* the corresponding CSI drivers support SELinux mount, kubelet mounts the volume with `-o context=system_u:object_r:container_file_t:s0:c10,c0`.
+* If `myclaim` is a RWOP volume (`Spec.AccessModes == ["ReadWriteOncePod']`)  *and* the corresponding CSI drivers support SELinux mount, kubelet files the blanks in the `seLinuxOptions` from the system defaults (`user: system_u`, `role: object_r`, `type: container_t` on Fedora based distros), translates them to a file label (`container_t` -> `container_file_t`) and mounts the volume with `-o cont ...
```
nit: fills the blank?

jsafrane replied:
Fixed

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -602,7 +606,12 @@ Drawbacks:
+  KCM often runs in a container and does not have access to `/etc/selinux` on the worker nodes.
+  As consequence, two labels that are equivalent from the SELinux point of view, may be reported as different, such as these two `seLinuxOptions` snippets: `{"type": "container_t", "level": "s0:c10,c0"}` and `{"level": "s0:c10,c1"}`.
+  `container_t` is the default type label for containers on Fedora, so kubelet is able to fill it in the `seLinuxOptions` when it is not set and see they're equivalent.
+  KCM does not know the default on nodes and treats these `seLinuxOptions` as different.
```
I thought we agreed that, if `level` of two containers are same but type is mismatched either ways (i.e one has type filled and other one doesn't), then we will consider these containers are incomparable and not emit warning. 

If we are going to emit warnings in these cases, then existing pods that could run fine will now trigger warnings?

jsafrane replied:
You're right! Updated (+ implementation will follow)

gnufied commented on keps/sig-storage/1710-selinux-relabeling/README.md:
```diff
@@ -668,10 +680,13 @@ No existing / new tests for volume mounting there.
-* GA:
+  * This phase signalizes that the feature is ready for real testing. Only non-breaking parts (`SELinuxChangePolicy`) are enabled by default.
+* GA of Phase 2 (`SELinuxChangePolicy` + `SELinuxMountReadWriteOncePod` are GA and locked to default):
   * All known issues fixed. Otherwise, we will GA Phase 1 only.
```
Can you clarify which phase we will be in when this is implemented? Will - `SELinuxChangePolicy` will have an material effect beyond allowing users to opt-out of mount-option for applying selinux labels?

What happens if `SELinuxMount` is enabled (because we are moving this to beta), will `SELinuxChangePolicy` will then have an material impact?

kannon92 commented:
PRR shadow:

I was reading the KEP and I think some updates for phase 1 may be needed.

https://github.com/kubernetes/enhancements/blob/fead564cd04409dc8f549ee655b3b5174c16fd29/keps/sig-storage/1710-selinux-relabeling/README.md#graduation-criteria

There is no mention of Phase 1 graduation. From the KEP issue, it sounds like you want to promote it to stable in 1.34.

kannon92 commented:
PRR shadow: looks good.
One item needs to be filled out.

This was tested manually before releasing SELinuxMountReadWriteOncePod enabled by default. It will be tested before SELinuxMount gets enabled by default.

kannon92 commented:
On a side note, what events are emitted in this case? The KEP mentions events but I didnt see what event would be emitted.

kannon92 commented:
And I think a prod-readiness should be updated as you are changing stage for one of the feature gates

https://github.com/kubernetes/enhancements/blob/master/keps/prod-readiness/sig-storage/1710.yaml

jsafrane commented:
> From the KEP issue, it sounds like you want to promote it to stable in 1.34

I updated the issue, 1.35 is the earliest and optimistic GA of SELinuxMountReadWriteOncePod + SELinuxChangePolicy. SELinuxMount at least 1 release later.

> There is no mention of Phase 1 graduation.

Good catch. I added "e2e tests implemented + green".

> And I think a prod-readiness should be updated as you are changing stage for one of the feature gates

I will do it in a subsequent PR, once the enhancement text is polished and merged.

jsafrane commented:
Ok, I added PRR request. We all three feature gates beta in 1.33.

gnufied commented:
from sig-storage and implementation POV. LGTM

/lgtm

jsafrane commented:
I tested upgrade + downgrade + upgrade, it seems working to me.

One interesting observation is that when SELinuxChangePolicy FG got from `true` to `false`, the API server did not clear `SELinuxChangePolicy` field in existing Deployments (this is expected), but then when new Pods were created from these Deployments, their `SELinuxChangePolicy` got cleared to `nil` on creation (again, expected, but not obvious).

After flipping `SELinuxChangePolicy` back to `true`, the Deployments kept their `SELinuxChangePolicy` and newly created Pods inherited it.


Details:
1. Run today's master (491a23f) with SELinuxMount && SELinuxChangePolicy off on Fedora 41, cri-o, SELinux enforcing, enabled SELinuxWarningController, single worker node (all the pods below run on the same node!)
SELinuxWarningController won't run, because it's behind SELinuxChangePolicy feature gate, but it's on KCM command line and we can just flip the feature gates in subsequent steps

2. Run 3 Deployments sharing the same RWO PVC in this sequence. Each has 1 replica.
    - `mydeployment-priv` with a Privileged Pod, no explicit SELinux label
    - `mydeployment-c1` with `seLinuxOpts{level:"c1,c0"}` and default SELinuxChangePolicy (nil)
    - `mydeployment-c2` with `seLinuxOpts{level:"c2,c0"}` and explicitly `Recursive` SELinuxChangePolicy

3. Set FG SELinuxChangePolicy: "true", SELinuxMount: "false" in KAS, KCM, scheduler and kubelet, restart them.

    - conflicting events appeared - every pod conflict with every other.

6. Set `SELinuxChangePolicy` of `c2` and `priv` to `Recursive`
    * `c1` still conflicts with `c2` and `priv`.
    * `c2` and `priv` do not conflict.

7. Set SELinuxChangePolicy: "true", SELinuxMount: "true" in KAS, KCM, scheduler and kubelet, restart them.

    * Depending on the order in which pods started, some Pods can get ContainerCreating (because `c1` wants `mount -o context=...c1,c0` and the other wants no `-o context`).
    * `c1` still conflicts with `c2` and `priv`.
    * `c2` and `priv` do not conflict.
   
8. Set SELinuxChangePolicy: "true", SELinuxMount: "false" in KAS, KCM, scheduler and kubelet, restart them.

    * All pods are back Running.
    * Same conflict events as before

9. Set SELinuxChangePolicy: "false", SELinuxMount: "false" in KAS, KCM, scheduler and kubelet, restart them.

    * All pods Running.
    * No conflicts (no SELinuxWarningController with `SELinuxChangePolicy=false`

10. Set SELinuxChangePolicy: "true", SELinuxMount: "true" in KAS, KCM, scheduler and kubelet, restart them.
    * (skipping SELinuxChangePolicy: "true", SELinuxMount: "false" to save time)
    * Depending on the order in which pods started, some Pods can get ContainerCreating.
    * `c1` still conflicts with `c2` and `priv`.
    * `c2` and `priv` do not conflict (= Deployments retained their `SELinuxChangePolicy` value from step 6).

deads2k commented:
PRR lgtm
/approve
